Code/GraphMol/Descriptors/BagOfBonds.cpp

```diff
@@ -0,0 +1,310 @@
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include <GraphMol/RDKitBase.h>
+#include <GraphMol/SmilesParse/SmilesParse.h>
+#include "BagOfBonds.h"
+#include "MolData3Ddescriptors.h"
+#include <Eigen/Dense>
+
+using namespace Eigen;
+
+
+struct BagMatrix {
+  VectorXd UCM;
+  std::vector<std::string > BagTag;
+};
+
+namespace RDKit {
+namespace Descriptors {
+namespace {
+
+
+VectorXd getUpperMat(MatrixXd mat) {
+  VectorXd res(mat.rows()*(mat.cols()+1)/2);
+  Index size = mat.rows();
+  Index offset = 0;
+  for(Index j=0; j<mat.cols(); ++j) {
+      res.segment(offset,size) = mat.col(j).tail(size);
+      offset += size;
+      size--;
+  }
+  return res;
+}
+
+std::list<std::pair<std::string, unsigned int> > EnumeratesAtomsPair(const RDKit::ROMol &mol, 
+    bool separateIsotopes, bool abbreviateHIsotopes) {
+  std::map<std::string, unsigned int> counts;
+  unsigned int nHs = 0;
+  std::string key;
+  const RDKit::PeriodicTable *table = RDKit::PeriodicTable::getTable();
+  for (RDKit::ROMol::ConstAtomIterator atomIt = mol.beginAtoms();
+       atomIt != mol.endAtoms(); ++atomIt) {
+    int atNum = (*atomIt)->getAtomicNum();
+    key = table->getElementSymbol(atNum);
+    if (separateIsotopes) {
+      unsigned int isotope = (*atomIt)->getIsotope();
+      if (abbreviateHIsotopes && atNum == 1 && (isotope == 2 || isotope == 3)) {
+        if (isotope == 2)
+          key = "D";
+        else
+          key = "T";
+      }
+    }
+    if (counts.find(key) != counts.end()) {
+      counts[key] += 1;
+    } else {
+      counts[key] = 1;
+    }
+    nHs += (*atomIt)->getTotalNumHs();
+  }
+
+  if (nHs) {
+    key = "H";
+    if (counts.find(key) != counts.end()) {
+      counts[key] += nHs;
+    } else {
+      counts[key] = nHs;
+    }
+  }
+
+  // create the Diagonal single atoms count
+  std::list<std::pair<std::string, unsigned int> > ks;
+  for (std::map<std::string, unsigned int>::const_iterator countIt = counts.begin();
```

Comment:
Again, these types of loops should be done using ranges:
`for(const auto &count : counts)`